qiskit/optimization/converters/optimization_problem_to_negative_value_oracle.py

```diff
@@ -0,0 +1,182 @@
+
+        Returns:
+            A state preparation operator A, an oracle O that recognizes negative numbers, and
+            a dictionary representation of the function coefficients, where the key -1 represents
+            the constant.
+        """
+
+        # get linear part of objective
+        linear_dict = problem.objective.get_linear()
+        linear_coeff = np.zeros(problem.variables.get_num())
+        for i, v in linear_dict.items():
+            linear_coeff[i] = v
+
+        # get quadratic part of objective
+        quadratic_dict = problem.objective.get_quadratic()
+        quadratic_coeff = {}
+        for i, jv in quadratic_dict.items():
+            for j, v in jv.items():
+                coeff = quadratic_coeff.get((j, i), 0)
+                if i <= j:
+                    # divide by 2 since problem considers xQx/2.
+                    quadratic_coeff[(i, j)] = v / 2 + coeff
+                else:
+                    quadratic_coeff[(j, i)] = v / 2 + coeff
+
+        constant = problem.objective.get_offset()
+
+        # Get circuit requirements from input.
+        self._num_key = len(linear_coeff)
+
+        # Get the function dictionary.
+        func = self._get_function(linear_coeff, quadratic_coeff, constant)
+        self._logger.info("Function: {}\n", func)
+
+        # Define state preparation operator A from function.
+        a_operator_circuit = self._build_operator(func)
+        a_operator = Custom(a_operator_circuit.width(), circuit=a_operator_circuit)
+
+        # Get registers from the A operator circuit.
+        reg_map = {}
+        for reg in a_operator_circuit.qregs:
+            reg_map[reg.name] = reg
+        key_val = reg_map["key_value"]
+
+        # Build negative value oracle O.
+        oracle_bit = QuantumRegister(1, "oracle")
+        oracle_circuit = QuantumCircuit(key_val, oracle_bit)
+        oracle_circuit.z(key_val[self._num_key])  # recognize negative values.
+        oracle = CustomCircuitOracle(variable_register=key_val,
+                                     output_register=oracle_bit,
+                                     circuit=oracle_circuit,
+                                     evaluate_classically_callback=self._evaluate_classically)
+
+        return a_operator, oracle, func
+
+    @staticmethod
+    def _get_function(linear: np.array, quadratic: np.array, constant: int) -> \
+            Dict[Union[int, Tuple[int, int]], int]:
+        """Convert the problem to a dictionary format."""
+        func = {-1: constant}
+        for i, v in enumerate(linear):
+            func[i] = v
+        for ij, v in quadratic.items():
+            i, j = ij
+            if i != j:
+                func[(i, j)] = int(quadratic[(i, j)])
+            else:
+                func[i] += v
+
+        return func
+
+    def _evaluate_classically(self, measurement):
+        # TODO: Typing for this method? Still not sure what it's used for. Required by Grover.
```

Comment:
No typing required since it's a private method
```suggestion
```

Reply:
My comment was more around what the method was meant to return. I don't think this has been implemented correctly, as I was struggling to find exactly what Grover was expecting with evaluate_classically.

Reply:
I think this should check whether the input is a solution, let me double check.